lib/Runtime/Library/LazyJSONString.h

```diff
@@ -0,0 +1,89 @@
+//-------------------------------------------------------------------------------------------------------
+// Copyright (C) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE.txt file in the project root for full license information.
+//-------------------------------------------------------------------------------------------------------
+
+#pragma once
+
+namespace Js
+{
+
+struct JSONProperty;
+struct JSONArray;
+
+enum class JSONContentType : uint8
+{
+    Array,
+    Object,
+    Undefined,
+    Null,
+    True,
+    False,
+    Number,
+    String
+};
+
+typedef DListCounted<JSONProperty, Recycler> JSONObject;
+
+struct JSONNumberData
+{
+    Field(Var) value;
+    Field(JavascriptString*) string;
+};
+
+struct JSONProperty
+{
+    Field(JSONContentType) type;
+    Field(const PropertyRecord*) propertyRecord;
+    Field(JavascriptString*) propertyName;
+    union
+    {
+        Field(JSONNumberData) numericValue;
+        Field(JavascriptString*) stringValue;
+        Field(JSONObject*) obj;
+        Field(JSONArray*) arr;
+    };
+
+    JSONProperty()
+    {
+        memset(this, 0, sizeof(JSONProperty));
+    }
+};
+
+struct JSONArray
+{
+    Field(uint32) length;
+    Field(JSONProperty) arr[];
+};
+
+class LazyJSONString : JavascriptString
+{
+private:
+    Field(charcount_t) gapLength;
+    Field(JSONProperty*) jsonContent;
+    Field(const char16*) gap;
+
+    DynamicObject* ParseObject(_In_ JSONObject* valueList) const;
```

Comment:
Nit: based on what these functions are doing, I'd prefer something like ReconstructVar/ReconstructObject/ReconstructArray for a slight improvement in readability

Reply:
Totally agree. We're not really parsing at all, are we?